src/utilities/theme/utils.ts

```diff
@@ -1,115 +1,779 @@
-import tokens from '@shopify/polaris-tokens';
-import {HSLColor} from '../color-types';
+import {HSLColor, HSLAColor} from '../color-types';
 import {colorToHsla, hslToString, hslToRgb} from '../color-transformers';
 import {isLight} from '../color-validation';
-import {constructColorName} from '../color-names';
-import {createLightColor} from '../color-manipulation';
-import {compose} from '../compose';
-
-import {Theme, ColorsToParse, ThemeVariant, ThemeColors} from './types';
-import {needsVariantList} from './config';
-
-export function setColors(theme: Theme | undefined): string[][] | undefined {
-  let colorPairs;
-  if (theme && theme.colors) {
-    Object.entries(theme.colors).forEach(([colorKey, pairs]) => {
-      const colorKeys = Object.keys(pairs);
-      if (colorKey === 'topBar' && colorKeys.length > 1) {
-        colorPairs = colorKeys.map((key: string) => {
-          const colors = (theme.colors as ThemeColors).topBar;
-          return [constructColorName(colorKey, key), colors[key]];
-        });
-      } else {
-        colorPairs = parseColors([colorKey, pairs]);
-      }
-    });
-  }
+import {CSSProperties, Theme} from './types';
 
-  return colorPairs;
+// TODO: Pull from polaris-tokens
+enum DefaultColor {
+  Surface = '#FAFAFA',
+  OnSurface = '#1F2225',
+  InteractiveNeutral = '#EAEAEB',
+  Interactive = '#0870D9',
+  Branded = '#008060',
+  Critical = '#E32727',
+  Warning = '#FFC453',
+  Highlight = '#59D0C2',
+  Success = '#008060',
 }
 
-export function needsVariant(name: string) {
-  return needsVariantList.indexOf(name) !== -1;
-}
+export function Colors(theme: Theme) {
+  const {colors = {}} = theme;
+
+  const isLightTheme = isLight(
+    hslToRgb(colorToHsla(
+      colors == null || colors.surface == null
+        ? DefaultColor.Surface
+        : colors.surface,
+    ) as HSLColor),
```

Comment:
Having to cast the result of colorToHsla() everywhere makes me think that return type is pretty crap.

Can you investigate what's going on there and make it so it always returns a type of HSLAColor or some other suitable type?

Reply:
This one is complicated, since the way we type color is a bit iffy, and will always result as any invalid color being considered a string.

```
function getColorType(color: string): ColorType {
  if (color.includes('#')) {
    return ColorType.Hex;
  } else if (color.includes('rgb')) {
    return ColorType.Rgb;
  } else if (color.includes('rgba')) {
    return ColorType.Rgba;
  } else if (color.includes('hsl')) {
    return ColorType.Hsl;
  } else if (color.includes('hsla')) {
    return ColorType.Hsla;
  } else {
    if (process.env.NODE_ENV === 'development') {
      /* eslint-disable-next-line no-console */
      console.warn('Accepted colors formats are: hex, rgb, rgba, hsl and hsla');
    }
    return ColorType.Default;
  }
}
```


```
export function colorToHsla(color: string): HSLAColor | string {
  const type: ColorType = getColorType(color);
  switch (type) {
    case ColorType.Hex:
      return hexToHsla(color);
    case ColorType.Rgb:
    case ColorType.Rgba:
      return rbgStringToHsla(color);
    case ColorType.Hsl:
    case ColorType.Hsla:
    default:
      return color;
  }
}
```

I think the best I can do here is add a typeguard that the color is not a string, since a string is essentially an invalid color. If this typeguard shows that our type is a string, it will warn to the console that the user provided an invalid color, and early return an empty object.

Thoughts on this? I'll push up the code shortly.

Reply:
If a string is an invalid color perhaps we should be throwing exceptions rather than letting that invalidness permeate through function calls?


```
export function strictColorToHsla(color: string): HSLAColor {
  const type: ColorType = getColorType(color);
  switch (type) {
    case ColorType.Hex:
      return hexToHsla(color);
    case ColorType.Rgb:
    case ColorType.Rgba:
      return rbgStringToHsla(color);
    case ColorType.Hsl:
    case ColorType.Hsla:
      return color
    default:
      throw new Error('hey that's not a valid color object');
  }
}

```